### PR TITLE
Allow filtering the query for list

### PR DESCRIPTION
**Pull request by amitaibu**
#124

### CODE REVIEW

File: plugins/restful/RestfulEntityBase.php [hunk 1]
```diff
@@ -627,10 +642,66 @@ public function getQueryForList() {
...
+   * query range.
+   *
+   * @param \EntityFieldQuery $query
+   *   The query object.
+   *
+   * @throws \RestfulBadRequestException
+   *
+   * @see \RestfulEntityBase::getQueryForList
+   */
+  protected function queryForListPagination(\EntityFieldQuery $query) {
```
Comment (e0ipso) [inline plugins/restful/RestfulEntityBase.php:691]: We now (#127) have a method to get the total count. That will allow us to know the number of pages in advance as well as avoiding the `$this->getRange() + 1` trick.

File: tests/modules/restful_test/plugins/restful/node/test_articles/1.2/RestfulTestArticlesResource__1_2.class.php [hunk 2]
```diff
@@ -18,19 +18,37 @@ public function getPublicFields() {
...
-    );
-
-    $public_fields['entity_reference_multiple'] = array(
-      'property' => 'entity_reference_multiple',
-      'resource' => array(
-        'article' => 'test_articles',
-      ),
-    );
+    // By checking that the field exists, we allow re-using this class on
+    // different tests, where different fields exist.
```
Comment (e0ipso) [inline tests/modules/restful_test/plugins/restful/node/test_articles/1.2/RestfulTestArticlesResource__1_2.class.php:22]: :+1:

**Comment by e0ipso** [comment]
I'm afraid to see this become a query builder. First adding the match for the field contents, then the _contains_, then the _NOT_, then operators with multiple filters, … I feel like we should come with a plan in advance for that.

What's the plan behind this? Do other API projects face this need, how do they solve them?

Is it for user defined filters?

**Comment by amitaibu** [comment]
> What's the plan behind this?

Here's my use case:
- "Company" content type (e.g. puma, adidas)
- "Item" content type (e.g. shirt, pants) that references a company.

`api/v1/items` -- gets all the items
`api/v1/items?filter[company]=1` -- get the items of a single company with ID 1

**Comment by amitaibu** [comment]
@mateu-aguilo-bosch ^^ 
makes sense to you?

**Comment by e0ipso** [comment]
@amitaibu sure. I think that we need to ask ourselves, _Is that the only filtering we'll want to allow?, filtering by exact match without condition composition_.

**Comment by amitaibu** [comment]
> Is that the only filtering we'll want to allow?, filtering by exact match without condition composition.

I've started building it with optionally allowing setting an [operator](https://github.com/Gizra/restful/pull/130/files#diff-3c241a6c3b4e93a88a7da3e91c3e9358R671)

We could also introduce `'allow filter' => TRUE` property, to allow switching it off.

What are your concerns?

**Comment by e0ipso** [comment]
My concerns are that we end up building a _Query Builder_ using EFQ and exposing a dev interface through the URL.
1. If we keep it simple, it makes sense this way.
2. If we go flexible we may need some advanced implementation for this (views does something like that, exposing filters and building queries).

**Comment by amitaibu** [comment]
The (still non-working) PR is currently more or less what I was planning. I'll finish it, and submit it for your review, and we'll figure out if it's too complex, ok?

**Comment by amitaibu** [comment]
This now includes #135.

**Comment by amitaibu** [comment]
@mateu-aguilo-bosch Tests are passing.

(We can wait for #135 and #127 to get it in first)

**Comment by amitaibu** [comment]
PR merged with #129

**Comment by amitaibu** [comment]
Merged.

**Pull request closed without merging**